tools/python/util/mobile_helpers/usability_checker.py

```diff
@@ -0,0 +1,558 @@
+                        domain_op = pieces[1]
+                        caveat = pieces[2]
+                        caveat = caveat.replace('<br/>', ' ')  # remove some HTML tags
+                        # skip lines that don't have the ':' which separates the domain and op
+                        # e.g. the table header will fail this check
+                        if ':' in domain_op:
+                            self._ops[domain_op] = caveat
+
+    def is_op_supported(self, node):
+        domain = node.domain if node.domain else 'ai.onnx'
+        domain_op = domain + ':' + node.op_type
+
+        is_supported = domain_op in self._ops
+        if is_supported:
+            self._ops_seen.add(domain_op)
+
+        return is_supported
+
+    def get_caveats(self):
+        caveats = []
+        for op in sorted(self._ops_seen):
+            caveat = self._ops[op]
+            if caveat:
+                caveats.append(f'{op}:{caveat}')
+
+        return caveats
+
+
+class PartitioningInfo:
+    class TryWithEP(IntEnum):
+        NO = 0,
+        MAYBE = 1,
+        YES = 2
+
+    def __init__(self):
+        self.num_nodes = -1  # main graph only
+        self.num_supported_nodes = -1
+        self.num_partitions = -1
+        self.num_nodes_in_subgraphs = -1  # nodes not covered as we don't currently handle subgraphs in nnapi/coreml
+        self.supported_ops_checker = None
+        self.supported_groups = []
+        self.unsupported_ops = set()
+        self.nodes_unsupported_due_to_op = -1
+        self.nodes_unsupported_due_to_dynamic_input = -1
+
+    def suitability(self):
+
+        # for now add up all the nodes. if there are subgraphs, the percentage of covered nodes will be reduced by all
+        # nodes in the subgraphs.
+        num_nodes = self.num_nodes + self.num_nodes_in_subgraphs
+
+        # semi-arbitrary choices that err on the side of MAYBE.
+        # having 1 partition is always preferred, but if that is small it may not be useful.
+        # having 2 partitions may be okay if they cover most nodes
+        # more than 2 partitions and the device copy cost is almost guaranteed to outweight the benefit of using the NPU
+        # NOTE: This assumes the EP is not CPU based and there is device copy overhead to consider
+        pct_supported = self.num_supported_nodes / num_nodes * 100
+        if self.num_partitions == 1:
+            if pct_supported > 75:
+                return PartitioningInfo.TryWithEP.YES
+            elif pct_supported > 50:
+                return PartitioningInfo.TryWithEP.MAYBE
+            else:
+                return PartitioningInfo.TryWithEP.NO
+
+        if self.num_partitions == 2:
+            if pct_supported > 75:
+                return PartitioningInfo.TryWithEP.MAYBE
+            else:
+                return PartitioningInfo.TryWithEP.NO
+
+        return PartitioningInfo.TryWithEP.NO
+
+    def analyze_info(self, logger: logging.Logger, ep_name: str):
```

Comment:
>analyze_info [](http://example.com/codeflow?start=8&length=12)

it looks like suitability() and analyze_info() do similar checks. perhaps they could be unified?

Reply:
suitability provides the yes/no/maybe. analyze info is for dumping all the logging info, which is overwhelming for a new user.

I did split out the yes/no/maybe as it was originally in analyze_info, so maybe analyze_info can be renamed.

Reply:
One benefit I see of combining them is having one place to maintain the suitability conditions instead of two. Possibly the logging level could be checked when generating the analysis output. I see that suitability and dump_analysis are called in such a way that the suitability result is output first. The analysis output could be saved to a string for later output if we want to display the overall suitability result first.